### PR TITLE
Modern setup.py: CMake-Driven, Multi-Dimensional

### PR DESCRIPTION
**Pull request by ax3l**
This adds a new, project-centric setup.py file. With this file, all dimensions (2D, 3D, RZ) of WarpX can be built and packaged at once, using the CMake build logic.

Build & install:
```bash
python3 -m pip wheel -v .
python3 -m pip install *whl
```
Via a source package:

```bash
python setup.py sdist --dist-dir .
pip wheel -v pywarpx-21.01.tar.gz
python3 -m pip install *whl
```
or directly with install:
```bash
python setup.py sdist --dist-dir .
pip install -v pywarpx-21.01.tar.gz
```

### To Do

- [x] add `requirements.txt` and `pyproject.toml`
- [x] write docs
- [x] test source package
- [x] test:
```bash
docker run -it -v ${PWD}:/warpx ubuntu:latest
```
```bash
apt update
apt upgrade

apt install -y build-essential cmake git libadios-dev libhdf5-dev python3-setuptools python3-pip python3-wheel
# apt install -y libopenmpi-dev libadios-openmpi-dev libhdf5-openmpi-dev
python3 -m pip install -U pip
```
```bash
WarpX_OPENPMD=ON python3 -m pip wheel -v .
python3 -m pip install *whl
```
- [x] run a serial PICMI setup

### Future

- potentially remove unused copies of "long-name" library names (cosmetic)

### CODE REVIEW

File: requirements.txt
Comment (dpgrote) [inline requirements.txt]: There probably should be a minimal version of numpy, but I don't know what it would be without testing.

File: setup.py
Comment (dpgrote) [inline setup.py:128]: Is there a reason that MPI is set to off (overriding the default)?

File: setup.py
Comment (dpgrote) [inline setup.py]: Is there a reason that YES is repeated?

File: setup.py
Comment (dpgrote) [inline setup.py]: Maybe add "plasma"

**Comment by ax3l** [comment]
LGTM note on `picmi.py`:
```
Class ElectrostaticFieldDiagnostic may not be initialized properly as method PICMI_ElectrostaticFieldDiagnostic.__init__ is not called from its __init__ method.
```

-> #1649

File: setup.py
Comment (ax3l) [inline setup.py]: VC discussion: could be an option.

File: setup.py
Comment (ax3l) [reply setup.py:128]: Discussed in VC: we disable dependencies that error out and fail the build. We will document how to build against MPI (also needs to disable binaries for transient dependencies like mpi4py).

File: setup.py
Comment (ax3l) [reply setup.py]: Ups, that's a mistake.

File: requirements.txt
Comment (ax3l) [inline requirements.txt:4]: PR open in: https://github.com/picmi-standard/picmi/pull/34

File: Docs/source/building/cmake.rst
Comment (ax3l) [inline Docs/source/building/cmake.rst:200]:
```suggestion
============================= ============================================ =======================================================

```

File: requirements.txt
Comment (ax3l) [reply requirements.txt:4]: does not hurt to keep this here until the next release of `picmistandard` is out

File: setup.py
Comment (dpgrote) [inline setup.py:192]: Why specify '<3.10'?

Comment (ax3l) [reply setup.py:192]: I usually do this because I want to test a new python release and then push a new package release that supports it first. This way users get a clean message that we have not yet released a Python 3.10 package that we know is good.

It's probably not needed here though (yet), but I know that e.g. middleware like pybind11 needs to be tested against a new CPython usually first. And then we want to build new wheels for a new Python release, too.

So I have no strong objection to keeping the upper range open, yet practically we release 1x a month and that's quick enough until users can upgrade (and there are many python version to fall back to).

File: requirements.txt
Comment (dpgrote) [inline requirements.txt:1]: Is there a reason that you use "~=" instead of ">="? The same for periodictable and scipy. (Sorry for the picky comments.)

Comment (ax3l) [reply requirements.txt:1]: Sure thing: it's because numpy and scipy use semantic versioning and it's unlikely we will work with a major release and its associated breaks before testing it.

This semantic here says: use `numpy>=1.15.0,<2.0.0`

Since I want to avoid being ping-ed by users on Slack because they got the two-hour-old numpy release and nothing works anymore, I rather like the pip dependency resolver  to load the last 1.* release of numpy for instance and update at our own speed in the next monthly release.
If both works, we just push ` numpy~=1.15.0,~=2.0.0` in the following release.

It's basically a way to channel releases and support at our own pace instead of being driven by (breaking) third-party releases (and their bugs).

File: .github/workflows/source/hasNonASCII
Comment (dpgrote) [inline .github/workflows/source/hasNonASCII:31]: I don't see any non-ascii in setup.py - why is it excluded?

Comment (ax3l) [reply .github/workflows/source/hasNonASCII:31]: I wrote Maxence Thévenet and Rémi Lehe in the author/maintainer strings :)
(The `é` is not part of ASCII.)

File: requirements.txt
Comment (dpgrote) [reply requirements.txt:1]: Ok, being conservative about releases is probably not a bad thing. Though it'll probably be a long time before we'll have NumPy version 2.

File: .github/workflows/source/hasNonASCII
Comment (dpgrote) [reply .github/workflows/source/hasNonASCII:31]: Ah. Their names are written with `e` in the source code, so maybe be consistent with that rather than have this one exception.

File: setup.py
Comment (dpgrote) [reply setup.py:192]: I thought that you might have had a specific reason to exclude 3.10 (like you knew about some upcoming API change that would break things).

I would be less conservative (and leave out the <3.10) but don't have a strong preference. Either way you'll get complaints from users, that 3.10 causes problems, or that 3.10 is not supported (even though it might work fine).

Comment (ax3l) [reply setup.py:192]: From experience, I know that CPython 3.8 had a performance change in memory buffers and CPython 3.9.0 had a memory error on macOS. Both of them were with pybind11 visible through openPMD-api, so I just kept the workflow :)

> Either way you'll get complaints from users, that 3.10 causes problems, or that 3.10 is not supported (even though it might work fine).

Just some subjective thoughts here on my experience with openPMD-api shippings:

True, but users that use conda-forge, brew, modules or their system python are "slow". For instance, most people even now [don't use Python 3.9 yet](https://github.com/openPMD/openPMD-api/issues/625#issuecomment-563032662). If we would be slow in releases (monthly is very fast), package managers like conda-forge would just automatically downgrade their python version and users might not even notice if they build a new WarpX environment that it's not cutting edge.

The users that go very cutting edge on CPython releases are also usually in the category that know how to downgrade and will not be blocked. The broad majority that just gets a breaking default install because a new breaking dependency was released on Saturday evening might be blocked by those - and also adds an urgency to bug reports that is harder to schedule for us maintainers (because they are blocked, need manual advise and we need to move faster) :)

File: requirements.txt
Comment (ax3l) [reply requirements.txt:1]: Yes, that's why I think it's quite save (I also try to avoid to over-constrain those)

File: .github/workflows/source/hasNonASCII
Comment (ax3l) [reply .github/workflows/source/hasNonASCII:31]: Good point, I just felt fancy because UTF8 works so well in Python :) @RemiLehe @MaxThevenet any preferences? This is the meta-data showing up on PyPI.

For C/C++ code we use ASCII only because some compilers and tools have sadly issues with non-ascii characters. (Just a matter of time until this will also become better.)

**Review by dpgrote — APPROVED** [review]
Looks good!

**Pull request merged**
